### ISSUE TITLE
broken tls support on Arch Linux

### ISSUE THREAD

**Issue by p5n** [couling/WebDAV-Daemon#7]
http works well, but https:
```
$ cadaver https://10.0.1.140:443
....
Username: user
Password:
Could not open collection:
Could not read status line: connection was closed by server
```

The same with nautilus and dolphin webdav clients.

Distro: ArchLinux

Versions:
webdav-daemon v1.1.r1.bc88ec6-2
libmicrohttpd 0.9.70-3

Config:
```
<?xml version="1.0" encoding="utf-8" ?>
<server-config xmlns="http://couling.me/webdavd">
	<server>
		<rap-binary>/usr/bin/rap</rap-binary>
		<static-response-dir>/usr/share/webdavd</static-response-dir>
		<chroot-path>~</chroot-path>
		<listen>
			<port>80</port>
			<encryption>none</encryption>
		</listen>
		<listen>
			<port>443</port>
			<encryption>ssl</encryption>
		</listen>
		<ssl-cert>
			<certificate>/etc/letsencrypt/live/.../cert.pem</certificate>
			<key>/etc/letsencrypt/live/.../privkey.pem</key>
		</ssl-cert>
	</server>
</server-config>
```

**Comment by couling** [comment]
Thanks for the report.

I think you are missing the intermediate certificates.  You either need to use the fullchain.pem instead of cert.pem or you need to use chain.pem as well as cert.pem.

I'll double check how I wrote the configuration for this. Off the top of my head I can't remember which way the TLS library works

**Comment by couling** [comment]
Thanks for your interest.  

Different software has different requirements but you must always use either of:

- cert.pem and chain.pem togeather
- fullchain.pem

Using just chain.pem on it's own is likely to fail because it doesn't contain the [intermediate certificate](https://en.wikipedia.org/wiki/Public_key_certificate#TLS/SSL_server_certificate) required by clients.  Different server software has different requirements so you must always check which you need.

For [webdavd you need cert.pem with chain.pem](https://github.com/couling/WebDAV-Daemon/blob/master/Configuration.md#ssl-cert):

    <ssl-cert>
        <certificate>/etc/letsencrypt/live/.../cert.pem</certificate>
        <chain>/etc/letsencrypt/live/.../chain.pem</chain>
        <key>/etc/letsencrypt/live/.../privkey.pem</key>
    </ssl-cert>

I'll close this issue for now.  If you still can't get it working, you can re-open and comment again.

**Comment by p5n** [comment]
I tried all variants with the same result. According to wireshark it connects TLSv1.3 successfully and  exchanges 4 application data packets after "Change Cipher Spec" one.

**Comment by p5n** [comment]
(I used wildcard let's encrypt cert if it can matter somehow)

**Comment by couling** [comment]
Okay I've never tested it on Arch Linux. I'm wondering about dependency trouble with the SSL library.  

What Arch Linux version are you running?

How did you install WebDAV-Daemon?

I'll see if I can recreate your issue.

**Comment by p5n** [comment]
current version of ArchLinux x86_64 (it is rolling release distro)
webdav installed from https://aur.archlinux.org/packages/webdav-daemon/

make package files: https://aur.archlinux.org/cgit/aur.git/tree/?h=webdav-daemon

main file: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=webdav-daemon

**Comment by p5n** [comment]
full output just in case:

```
$ cadaver http://10.0.1.140:1080
Authentication required for My Server on server `10.0.1.140':
Username: ...
Password: 
dav:/> ls
Listing collection `/': succeeded.
Coll:   test                                   0  мая 23 18:04
        qwe.txt                                0  мая 23 18:02
dav:/> 
Connection to `10.0.1.140' closed.
```
```
$ cadaver https://10.0.1.140:1443
WARNING: Untrusted server certificate presented for `*.mydomain.com':
Certificate was issued to hostname `*.mydomain.com' rather than `10.0.1.140'
This connection could have been intercepted.
Issued to: mydomain.com
Issued by: Let's Encrypt, US
Certificate is valid from Fri, 10 Apr 2020 10:04:51 GMT to Thu, 09 Jul 2020 10:04:51 GMT
Do you wish to accept the certificate? (y/n) y
Authentication required for My Server on server `10.0.1.140':
Username: .....
Password: 
Could not open collection:
Could not read status line: connection was closed by server
dav:/?

**Comment by montvid** [comment]
So does this code work with https? I am looking for a webdav solution that has two users one with read access only and the other with all access serving on the internet on port 443.

**Comment by couling** [comment]
@montvid Yes it supports SSL/TLS.  This issue was with building it for Arch Linux - a newer version of a supporting library (`libmicrohttpd`) broke the code.  As far as I'm aware this only affects Arch Linux

**Comment by couling** [comment]
For reference I reported this as a bug against `libmicrohttpd` [here](https://mail.gnu.org/archive/html/libmicrohttpd/2020-06/msg00006.html) with response [here](https://mail.gnu.org/archive/html/libmicrohttpd/2020-06/msg00007.html).

Unfortunately I then lost all my free time and couldn't get back to it.

**Comment by p5n** [comment]
It looks fixed with `libmicrohttpd 0.9.73-1` and latest webdav-daemon.
Thank you!
